org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/ClasspathAttributeDelta.java

```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Simeon Andreev and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.jdt.internal.core;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+
+import org.eclipse.core.runtime.Assert;
+import org.eclipse.jdt.core.IClasspathAttribute;
+import org.eclipse.jdt.core.IClasspathAttributeDelta;
+import org.eclipse.jdt.core.IClasspathEntry;
+
+/**
+ * @see IClasspathAttributeDelta
+ */
+public class ClasspathAttributeDelta implements IClasspathAttributeDelta {
+
+	private static final String NO_VALUE = ""; //$NON-NLS-1$
+
+	private final int kind;
+	private final String attributeName;
+	private final String attributeValue;
+
+	private ClasspathAttributeDelta(int kind, String attributeName, String attributeValue) {
+		Assert.isTrue(kind == ADDED || kind == REMOVED || kind == CHANGED, "Unexpected delta type: " + kind); //$NON-NLS-1$
+		this.kind = kind;
+		this.attributeName = attributeName;
+		this.attributeValue = attributeValue;
+	}
+
+	@Override
+	public int getKind() {
+		return this.kind;
+	}
+
+	@Override
+	public String getAttributeName() {
+		return this.attributeName;
+	}
+
+	@Override
+	public String getAttributeValue() {
+		return this.attributeValue;
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder buffer = new StringBuilder();
+		buffer.append("attribute name="); //$NON-NLS-1$
+		buffer.append(getAttributeName());
+		int deltaKind = getKind();
+		if (deltaKind != REMOVED) {
+			buffer.append(", value="); //$NON-NLS-1$
+			buffer.append(getAttributeValue());
+		}
+		buffer.append("["); //$NON-NLS-1$
+		switch (deltaKind) {
+			case ADDED :
+				buffer.append('+');
+				break;
+			case REMOVED :
+				buffer.append('-');
```

Comment:
i see removed changes for javadoc and module, although i just toggle test for a single jar on classpath

```
<module:jdk.net>[*]: {ANNOTATIONS | ATTRIBUTES}
	attribute name=javadoc_location[-]
	attribute name=module[-]
```